scalar/scalar.go

```diff
@@ -95,27 +148,83 @@ func AppendToBuilder(bldr array.Builder, s Scalar) {
 		bldr.(*array.BinaryBuilder).Append(s.(*LargeBinary).Value)
 	case arrow.STRING:
 		bldr.(*array.StringBuilder).Append(s.(*String).Value)
+	case arrow.LARGE_STRING:
+		bldr.(*array.LargeStringBuilder).Append(s.(*LargeString).s.Value)
 	case arrow.INT64:
-		bldr.(*array.Int64Builder).Append(s.(*Int64).Value)
+		bldr.(*array.Int64Builder).Append(s.(*Int).Value)
+	case arrow.INT32:
+		bldr.(*array.Int32Builder).Append(int32(s.(*Int).Value))
+	case arrow.INT16:
+		bldr.(*array.Int16Builder).Append(int16(s.(*Int).Value))
+	case arrow.INT8:
+		bldr.(*array.Int8Builder).Append(int8(s.(*Int).Value))
 	case arrow.UINT64:
-		bldr.(*array.Uint64Builder).Append(s.(*Uint64).Value)
+		bldr.(*array.Uint64Builder).Append(s.(*Uint).Value)
+	case arrow.UINT32:
+		bldr.(*array.Uint32Builder).Append(uint32(s.(*Uint).Value))
+	case arrow.UINT16:
+		bldr.(*array.Uint16Builder).Append(uint16(s.(*Uint).Value))
+	case arrow.UINT8:
+		bldr.(*array.Uint8Builder).Append(uint8(s.(*Uint).Value))
+	case arrow.FLOAT16:
+		bldr.(*array.Float16Builder).Append(float16.New(float32(s.(*Float).Value)))
 	case arrow.FLOAT32:
-		bldr.(*array.Float32Builder).Append(s.(*Float32).Value)
+		bldr.(*array.Float32Builder).Append(float32(s.(*Float).Value))
 	case arrow.FLOAT64:
-		bldr.(*array.Float64Builder).Append(s.(*Float64).Value)
+		bldr.(*array.Float64Builder).Append(s.(*Float).Value)
 	case arrow.BOOL:
 		bldr.(*array.BooleanBuilder).Append(s.(*Bool).Value)
 	case arrow.TIMESTAMP:
 		bldr.(*array.TimestampBuilder).Append(arrow.Timestamp(s.(*Timestamp).Value.UnixMicro()))
+	case arrow.DURATION:
+		bldr.(*array.DurationBuilder).Append(arrow.Duration(s.(*Duration).Value))
+	case arrow.DATE32:
+		bldr.(*array.Date32Builder).Append(s.(*Date32).Value)
+	case arrow.DATE64:
+		bldr.(*array.Date64Builder).Append(s.(*Date64).Value)
+	case arrow.TIME32:
+		bldr.(*array.Time32Builder).Append(arrow.Time32(int32(s.(*Time).Value)))
+	case arrow.TIME64:
+		bldr.(*array.Time64Builder).Append(arrow.Time64(s.(*Time).Value))
+	case arrow.INTERVAL_MONTHS:
+		bldr.(*array.MonthIntervalBuilder).Append(arrow.MonthInterval(int32(s.(*MonthInterval).Value)))
+	case arrow.INTERVAL_DAY_TIME:
+		bldr.(*array.DayTimeIntervalBuilder).Append(s.(*DayTimeInterval).Value)
+	case arrow.INTERVAL_MONTH_DAY_NANO:
+		bldr.(*array.MonthDayNanoIntervalBuilder).Append(s.(*MonthDayNanoInterval).Value)
+	case arrow.DECIMAL128:
+		bldr.(*array.Decimal128Builder).Append(s.(*Decimal128).Value)
+	case arrow.DECIMAL256:
+		bldr.(*array.Decimal256Builder).Append(s.(*Decimal256).Value)
+	case arrow.STRUCT:
+		sb := bldr.(*array.StructBuilder)
+		sb.Append(true)
+
+		v := s.(*Struct).Value
+		m := v.(map[string]any)
+		st := sb.Type().(*arrow.StructType)
+
+		for i, f := range st.Fields() {
+			sc := NewScalar(sb.FieldBuilder(i).Type())
+
+			if sv, ok := m[f.Name]; ok {
+				if err := sc.Set(sv); err != nil {
+					panic(err)
+				}
+				delete(m, f.Name)
```

Comment:
This will mutate the passed map value.
```go
names := make(map[string]struct{}, len(m))
for k := range m {
	names[k] = struct{}{}
}

...
if sv, ok := m[f.Name]; ok {
	delete(names, f.Name)
}
...
if len(names) > 0 {
	panic(fmt.Errorf("struct has extra fields: %+v", m))
}
```